### PR TITLE
Fix data type conversion during unmarshal

### PR DESCRIPTION
**Pull request by singhashmeet**
Javascript has only 1 64 bit data type ie float64 
hence any 64 bit variable is by default translated to a float64
custom unmarshal and clean operation fixes that issue

### CODE REVIEW

File: pkg/specs/property.go [hunk 1]
```diff
@@ -83,6 +85,43 @@ type Property struct {
...
+			if !ok {
+				prop.Default = int64(prop.Default.(float64))
+			}
+		case types.Uint64, types.Fixed64:
+			_, ok := prop.Default.(uint64)
+			if !ok {
+				prop.Default = uint64(prop.Default.(float64))
+			}
+		}
+	}
```
Comment (jeroenrinzema) [inline pkg/specs/property.go:122]: Please also handle 32 bit types

**Review by jeroenrinzema — CHANGES_REQUESTED** [review]

**Comment by jeroenrinzema** [comment]
Please include some test cases for this PR to improve the test coverage.

File: pkg/specs/property.go (hunk 1 above)
Comment (singhashmeet) [reply pkg/specs/property.go:122]: are 32 bits also converted to float64 or float32 ?

Comment (jeroenrinzema) [reply pkg/specs/property.go:122]: float32, check out the protobuf types conversion for all the available types and their go representation: https://developers.google.com/protocol-buffers/docs/proto3#scalar

File: pkg/specs/property_test.go [hunk 2]
```diff
@@ -8,6 +8,17 @@ import (
...
 
+func TestPropertyUnmarshal(t *testing.T) {
+	payload := `{"property":{"position":1,"name":"com.semaphore.products.Product","type":"message","label":"optional","nested":{"pcode":{"position":2,"name":"pcode","path":"pcode","type":"int64","label":"optional"},"pcode2":{"position":2,"name":"pcode2","path":"pcode2","type":"uint64","label":"optional"},"product":{"position":1,"name":"product","path":"product","type":"string","label":"optional"}}}}`
+
+	prop := Property{}
+	err := prop.UnmarshalJSON([]byte(payload))
+
+	if err != nil {
+		t.Errorf("unexpected error %+v", err)
+	}
```
Comment (jeroenrinzema) [inline pkg/specs/property_test.go:19]: please check whether the correct value and type is set.

**Review by jeroenrinzema — CHANGES_REQUESTED** [review]

File: pkg/specs/property.go (hunk 1 above)
Comment (singhashmeet) [reply pkg/specs/property.go]: Added

File: pkg/specs/property.go [hunk 3]
```diff
@@ -83,6 +85,53 @@ type Property struct {
...
+			_, ok := prop.Default.(uint64)
+			if !ok {
+				prop.Default = uint64(prop.Default.(float64))
+			}
+		case types.Int32, types.Sint32, types.Sfixed32:
+			_, ok := prop.Default.(int32)
+			if !ok {
+				prop.Default = int32(prop.Default.(float64))
+			}
+		case types.Uint32:
```
Comment (jeroenrinzema) [inline pkg/specs/property.go:126]: your missing `types.Fixed32` atm

File: pkg/specs/property_test.go [hunk 4]
```diff
@@ -1,13 +1,49 @@
...
+	payload := "non json string"
+	prop := Property{}
+	err := prop.UnmarshalJSON([]byte(payload))
+	if err == nil {
+		t.Error("expected error got nil")
+	}
+}
+func TestPropertyUnmarshal(t *testing.T) {
+	payload := `{"position":1,"name":"com.semaphore.products.Product","type":"message","label":"optional","nested":{"pcode":{"position":2,"name":"pcode","path":"pcode","type":"int64","label":"optional", "default": 100},"pcode2":{"position":2,"name":"pcode2","path":"pcode2","type":"uint64","label":"optional", "default":20},"pcode3":{"position":2,"name":"pcode3","path":"pcode3","type":"uint32","label":"optional", "default":20},"pcode4":{"position":2,"name":"pcode4","path":"pcode4","type":"int32","label":"optional", "default":20},"product":{"position":1,"name":"product","path":"product","type":"string","label":"optional"}}}`
+
```
Comment (jeroenrinzema) [inline pkg/specs/property_test.go:22]: It is best to split up these tests and include all possible type combinations: 
```		case types.Int64, types.Sint64, types.Sfixed64:
			_, ok := prop.Default.(int64)
			if !ok {
				prop.Default = int64(prop.Default.(float64))
			}
		case types.Uint64, types.Fixed64:
			_, ok := prop.Default.(uint64)
			if !ok {
				prop.Default = uint64(prop.Default.(float64))
			}
		case types.Int32, types.Sint32, types.Sfixed32:
			_, ok := prop.Default.(int32)
			if !ok {
				prop.Default = int32(prop.Default.(float64))
			}
		case types.Uint32, types.Fixed32:```

**Review by jeroenrinzema — CHANGES_REQUESTED** [review]

**Comment by jeroenrinzema** [comment]
Almost there 😉

File: pkg/specs/property.go (hunk 3 above)
Comment (singhashmeet) [reply pkg/specs/property.go:126]: fixed in line 116

**Review by jeroenrinzema — APPROVED** [review]

**Pull request merged**